### PR TITLE
EES-4646 - Adding backend validation to stop the creation, or updating, of a release type to Experimental statistics

### PR DESCRIPTION
**Pull request by jack-hive**

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs [hunk 1]
```diff
@@ -48,6 +49,36 @@ public class ReleaseServiceTests
...
+        public async Task CreateRelease_ReleaseTypeExperimentalStatistics_ReturnsValidationActionResult()
+        {
+            var releaseCreateRequest = new ReleaseCreateRequest
+            {
+                Type = ReleaseType.ExperimentalStatistics,
+            };
+
+            ReleaseService releaseService = BuildReleaseService(Mock.Of<ContentDbContext>());
+
+            Either<ActionResult, ReleaseViewModel> result = await releaseService.CreateRelease(releaseCreateRequest);
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs:62]: Hey there, we'd normally shorten this to:

`var result = await releaseService.CreateRelease(releaseCreateRequest);`

Similarly, `var releaseService = BuildReleaseService(Mock.Of<ContentDbContext>());`

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs [hunk 2]
```diff
@@ -48,6 +49,36 @@ public class ReleaseServiceTests
...
+        public async Task UpdateRelease_ReleaseTypeExperimentalStatistics_ReturnsValidationActionResult()
+        {
+            var releaseUpdateRequest = new ReleaseUpdateRequest
+            {
+                Type = ReleaseType.ExperimentalStatistics,
+            };
+
+            ReleaseService releaseService = BuildReleaseService(Mock.Of<ContentDbContext>());
+
+            Either<ActionResult, ReleaseViewModel> result = await releaseService.UpdateRelease(It.IsAny<Guid>(), releaseUpdateRequest);
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs:77]: As per other comment regarding shortening.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Validators/ReleaseRequestValidators.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using Microsoft.AspNetCore.Mvc;
+using GovUk.Education.ExploreEducationStatistics.Admin.Requests;
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Validators.ValidationErrorMessages;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Validators.ValidationUtils;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Validators;
+
+public static class ReleaseCreateRequestValidator
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Validators/ReleaseRequestValidators.cs:10]: It's nice to have the validation very clearly available like this :)

Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Validators/ReleaseRequestValidators.cs:10]: Just wondering, and it doesn't have to be part of this PR, but if it's possible to use validation attributes on the Request objects themselves to achieve this same behaviour?  Might be one to put on a back burner until we get round to getting FluentValidation into the mix properly though.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Model/EitherTest.cs [hunk 4]
```diff
@@ -150,6 +150,50 @@ public void OnSuccess_NoArg_Left_PriorFailure()
             Assert.Empty(results);
         }
 
+        [Fact]
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Model/EitherTest.cs:153]: Nice!

File: src/GovUk.Education.ExploreEducationStatistics.Common/Model/Either.cs [hunk 5]
```diff
@@ -35,6 +35,16 @@ public Either(TR right)
 
         public Either<TL, T> OnSuccess<T>(Func<TR, T> func) => Map(func);
 
+        public async Task<Either<TL, T>> OnSuccess<T>(Func<Task<Either<TL, T>>> func)
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Common/Model/Either.cs:38]: Haha, I think we've both added the same extension method in separate PRs - your test coverage is better though :)

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs (hunk 1 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs]: Done :) 

We have discussed this in the backend style-guide session and have agreed that it is a coding standard we have adopted and will take moving forwards

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs (hunk 2 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs]: done :)

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Model/EitherTest.cs (hunk 4 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Model/EitherTest.cs:153]: thanks :D

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Validators/ReleaseRequestValidators.cs (hunk 3 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Validators/ReleaseRequestValidators.cs:10]: I can't find one, but maybe there is? Would be interested to know if anyone knows about it

I discussed putting these validators in their own file/class with Ben. I suggested that it might be a stepping stone to moving towards using FluentValidation. We have these dedicated classes for validating which we can then hook into when we start to use FluentValidation - we just make the classes extend `AbstractValidator<T>` and make any necessary changes. 

Would be good to get some wider opinions on this too.

This is what made it a necessity to add the new `Either` extension method which validates the request object first, up-front, before making any DB calls to check for Entity existence etc...

**Review by duncan-at-hiveit — APPROVED** [review]
